gateway/src/resty/synchronization.lua

```diff
@@ -53,4 +53,28 @@ function _M:release(key)
   semaphores[key] = nil
 end
 
+--- run a new function, callback using locks on the given key
+-- @tparam string key: key for the semaphore
+-- @tparam int timeout: timeout for getting the lock before raise an error.
+-- @tparam function callback: function to execute if the lock is adquired correctly
+function _M:run(key, timeout, callback, ...)
+  local sema = self:acquire(key)
+
+  local ok, err = sema:wait(timeout)
+  if not ok then
+    ngx.log(ngx.WARN, 'failed to acquire lock on key: ', key, ' error: ', err)
+    return false, nil
+  end
+
+  local task = safe_task.new(callback)
+  local ret, result = task:execute(...)
+
+  if ok then
```

Comment:
Can we name this `lock_acquired` or similar? I think that would make it perfectly clear what this refers to.